### PR TITLE
Front End Components for Job Poster Builder - Impact

### PR DESCRIPTION
**Pull request by xander-miller**
Front End Components for #916

Still more work needs to be done on #916

### CODE REVIEW

**Comment by xander-miller** [comment]
Adds front end elements for Job Poster Builder Impact Step

File: resources/assets/js/components/JobBuilderImpact/ImpactForm.tsx [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+              data-c-button="solid(c1)"
+              data-c-dialog-action="open"
+              data-c-dialog-id="example-dialog-01"
+              data-c-radius="rounded"
+              disabled={isSubmitting}
+              form="form"
+              type="submit"
+            >
+              <FormattedMessage
+                id="button.next"
```
Comment (tristan-orourke) [inline resources/assets/js/components/JobBuilderImpact/ImpactForm.tsx]: A given formattedMessage id should only ever exist in one place. "button.next" will probably be used somewhere else, so we should either change this id to something page-specific (eg jobBuilder.impact.submitButton) or extract it to a common file where it can be reused elsewhere.

File: resources/assets/js/components/JobBuilderImpact/JobBuilderImpact.tsx [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+          <FormattedMessage
+            id="jobBuilder.impact.header.department"
+            defaultMessage="How our department makes an impact:"
+            description="Header of Department Impact Section on Job Poster Builder Impact Step"
+          />
+        </p>
+        {/* <!-- The p tag below is where the dynamic department text goes (I used text from an old job poster as placeholder, but the real list of data is in the issue. --> */}
+        <p data-c-margin="bottom(double)">
+          {department !== undefined &&
+            intl.formatMessage(departmentImpactStatements[department])}
```
Comment (tristan-orourke) [inline resources/assets/js/components/JobBuilderImpact/JobBuilderImpact.tsx]: Department displayed should be based on job.department_id.

Comment (tristan-orourke) [reply resources/assets/js/components/JobBuilderImpact/JobBuilderImpact.tsx]: I know the current implementation allows for easily changing the department in storybook, but it doesn't make sense outside of storybook.

File: resources/assets/js/components/Forms/TextArea.tsx [hunk 3]
```diff
@@ -12,6 +12,10 @@ export interface TextAreaProps {
   errorText?: string;
   onChange: (event: React.ChangeEvent<HTMLTextAreaElement>) => void;
   onBlur?: (event: React.ChangeEvent<HTMLTextAreaElement>) => void;
+
+  // formik
+  field?: any;
```
Comment (Khristophor) [inline resources/assets/js/components/Forms/TextArea.tsx]: @yonikid15 's PR #1052 will affect this, but we decided to keep Formik-specific components in the `Form/` directory and leave the existing input components in the `components/` directory.

File: resources/assets/js/translations/extractedMessages/resources/assets/js/components/ApplicationReview/ReviewApplications.json [hunk 4]
```diff
@@ -9,4 +9,4 @@
     "description": "Welcome header on app main page",
     "defaultMessage": "{dayCount, plural, =0 {No Days} one {# Day} other {# Days}} Since Close"
   }
-]
+]
```
Comment (Khristophor) [inline resources/assets/js/translations/extractedMessages/resources/assets/js/components/ApplicationReview/ReviewApplications.json]: Not sure how this is happening, since we have `files.insertFinalNewline` turned on in the `.vscode` settings, but we shouldn't be committing files that have no changes or are unrelated to the current issue.

File: resources/assets/js/translations/extractedMessages/resources/assets/js/components/AssessmentPlan/AssessmentPlanTable.json [hunk 5]
```diff
@@ -27,7 +27,7 @@
   {
     "id": "assessmentPlan.summary.description",
     "description": "Description of Assessment Plan Summary Section",
-    "defaultMessage": "This is a summary of the work you've done above. You'll find\n      each assessment accompanied by a consolidated list of the essential and\n      asset skills attached to it."
+    "defaultMessage": "This is a summary of the work you've done above. You'll find\r\n      each assessment accompanied by a consolidated list of the essential and\r\n      asset skills attached to it."
```
Comment (Khristophor) [inline resources/assets/js/translations/extractedMessages/resources/assets/js/components/AssessmentPlan/AssessmentPlanTable.json]: Same as above, if code is updated from an automated tool like translation manager, only the bits that apply to this story should be staged and the rest should go up in another PR.

File: resources/assets/js/stories/JobBuilder/JobBuilderImpact.stories.tsx [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+
+// const selectLang = ;
+
+stories
+  .add(
+    "Job Builder Impact",
+    (): React.ReactElement => (
+      <JobBuilderImpact
+        department={select("Department", deptOptions, "treasuryBoard")}
+        job={null}
```
[the remark this answers is not in the document]
Comment (tristan-orourke) [reply resources/assets/js/stories/JobBuilder/JobBuilderImpact.stories.tsx]: I like having to pass in `null`. I find it more explicit and easier to work with.

File: resources/assets/js/translations/locales/fr.json [hunk 7]
```diff
@@ -138,8 +178,39 @@
...
   "jobDetails.titleLabel": "What is the job title?",
   "jobDetails.titlePlaceholder": "e.g. Product Designer",
+  "languageRequirment.bilingualAdvanced": "Bilingual - Advanced",
+  "languageRequirment.bilingualIntermediate": "Bilingual - Intermediate",
+  "languageRequirment.english": "English essential",
+  "languageRequirment.englishOrFrench": "English or French",
+  "languageRequirment.french": "French essential",
   "priorityStatus.priority": "Priorité",
   "priorityStatus.priorityLogoTitle": "icône pour candidat prioritaire",
+  "province.ab": "Alberta",
```
[the remark this answers is not in the document]
Comment (tristan-orourke) [reply resources/assets/js/translations/locales/fr.json]: Yes, but I had already added provinces to `lookupConstants` at that point.

File: resources/assets/js/components.tsx [hunk 8]
```diff
@@ -13,3 +13,4 @@ require("./components/ApplicantSkills/SkillsInfoModal");
 require("./components/JobDetails/JobDetails");
 require("./components/JobBuilderIntro/JobBuilderIntro");
 
+// import ReviewApplications from './components/ReviewApplications';
```
[the remark this answers is not in the document]
Comment (xander-miller) [reply resources/assets/js/components.tsx]: Done

File: resources/assets/js/components/JobBuilderImpact/ImpactForm.tsx [hunk 9]
```diff
@@ -0,0 +1,186 @@
...
+      intl.formatMessage(validationMessages.required),
+    ),
+  });
+  return (
+    <Formik
+      initialValues={initialValues}
+      validationSchema={validationSchema}
+      onSubmit={(values, actions): void => {
+        // The following only triggers after validations pass
+        // handleSubmit(values);
```
[the remark this answers is not in the document]
Comment (xander-miller) [reply resources/assets/js/components/JobBuilderImpact/ImpactForm.tsx]: It's not connected to redux yet and there is no definition for values.

File: resources/assets/js/components/JobBuilderImpact/ImpactForm.tsx [hunk 10]
```diff
@@ -0,0 +1,186 @@
...
+interface ImpactFormProps {
+  job: Job | null;
+  handleSubmit: (job: Job) => void;
+}
+
+const ImpactForm: React.FunctionComponent<
+  ImpactFormProps & InjectedIntlProps
+> = ({ intl, job, handleSubmit }): React.ReactElement => {
+  const initialValues: ImpactFormValues = {
+    teamImpact: job ? job[intl.locale].team_impact : "",
```
[the remark this answers is not in the document]
Comment (xander-miller) [reply resources/assets/js/components/JobBuilderImpact/ImpactForm.tsx]: makes sense. Done.

File: resources/assets/js/components/JobBuilderImpact/ImpactForm.tsx [hunk 11]
```diff
@@ -0,0 +1,186 @@
...
+              It doesn’t matter if your work is direct to citizens or back office,
+              innovative or maintenance, top priority or ongoing. Describe how it
+              contributes to making Canada better the way you would to someone who
+              knows nothing about your work."
+                description="Body of Job Poster Builder Team Impact Section"
+              />
+            </p>
+            <span>
+              <FormattedMessage
+                id="formField.required"
```
[the remark this answers is not in the document]
Comment (xander-miller) [reply resources/assets/js/components/JobBuilderImpact/ImpactForm.tsx]: In an earlier version of the component it wasn't handled, the new version of the component isn't localized. I'll fix it.

File: resources/assets/js/stories/JobBuilder/JobBuilderImpact.stories.tsx (hunk 6 above)
[the remark this answers is not in the document]
Comment (Khristophor) [reply resources/assets/js/stories/JobBuilder/JobBuilderImpact.stories.tsx]: I guess @tristan-orourke and I differ on our opinions on this, so ignore me for now.

File: resources/assets/js/components/JobBuilderImpact/ImpactForm.tsx (hunk 11 above)
[the remark this answers is not in the document]
Comment (xander-miller) [reply resources/assets/js/components/JobBuilderImpact/ImpactForm.tsx]: Fixed.

File: resources/assets/js/components/JobBuilderImpact/ImpactForm.tsx (hunk 1 above)
Comment (xander-miller) [reply resources/assets/js/components/JobBuilderImpact/ImpactForm.tsx]: The `button.next` translation id is a common translation that could be used in other places. I don't think we have to extract out the whole formatted message so it is only used once.

File: resources/assets/js/components/JobBuilderImpact/JobBuilderImpact.tsx (hunk 2 above)
Comment (xander-miller) [reply resources/assets/js/components/JobBuilderImpact/JobBuilderImpact.tsx]: The idea was to build the front end without linking it to jobs yet. The department impact statement should really come from the database, as the list of partner departments will change with every year and with every election.

File: resources/assets/js/components/JobBuilderImpact/ImpactForm.tsx [hunk 12]
```diff
@@ -0,0 +1,186 @@
...
+            <span>
+              <FormattedMessage
+                id="formField.error"
+                defaultMessage="This input has an error."
+                description="Flag text for validation error"
+              />
+            </span>
+          </div>
+          <div data-c-alignment="centre" data-c-grid-item="base(1of1)">
+            {/* <!-- Modal trigger, same as last step. --> */}
```
[the remark this answers is not in the document]
Comment (xander-miller) [reply resources/assets/js/components/JobBuilderImpact/ImpactForm.tsx]: Probably, I will work on adding it.

File: resources/assets/js/stories/JobBuilder/JobBuilderImpact.stories.tsx [hunk 13]
```diff
@@ -0,0 +1,53 @@
...
+
+const stories = storiesOf("Job Builder - Impact", module)
+  .addDecorator(withInfo)
+  .addDecorator(withKnobs)
+  .addDecorator(withIntl);
+
+const deptOptions = {
+  "Treasury Board": "treasuryBoard",
+  "Natural Resources": "naturalResources",
+  Transport: "transport",
```
[the remark this answers is not in the document]
Comment (xander-miller) [reply resources/assets/js/stories/JobBuilder/JobBuilderImpact.stories.tsx]: The linter removes the quotes because the key is a single word.

File: resources/assets/js/stories/JobBuilder/JobBuilderImpact.stories.tsx [hunk 14]
```diff
@@ -0,0 +1,53 @@
...
+  "Global Affairs": "globalAffairs",
+  "Boarder Service Agency": "boarderServicesAgency",
+  "Innovation and Science": "innovationScience",
+  "Public Service Procurement": "publicServiceAndProcurement",
+  "Department National Defence": "departmentNationalDefence",
+  "Shared Services Canada": "sharedServicesCanada",
+  "Health Canada": "healthCanada",
+};
+
+// const selectLang = ;
```
[the remark this answers is not in the document]
Comment (xander-miller) [reply resources/assets/js/stories/JobBuilder/JobBuilderImpact.stories.tsx]: Done.

File: resources/assets/js/translations/locales/fr.json (hunk 7 above)
[the remark this answers is not in the document]
Comment (xander-miller) [reply resources/assets/js/translations/locales/fr.json]: Yes, I think the department impact statements are more critical to move into the database, as they are more likely to change than the provinces. If someone works with me on moving the department information perhaps I could then do the provinces myself.

File: resources/assets/js/components/Forms/TextArea.tsx (hunk 3 above)
Comment (xander-miller) [reply resources/assets/js/components/Forms/TextArea.tsx]: Yes, I want to get this code in. I will happily help refactor for Yoni's upcoming change.

File: resources/assets/js/translations/locales/fr.json (hunk 7 above)
[the remark this answers is not in the document]
Comment (xander-miller) [reply resources/assets/js/translations/locales/fr.json]: Grant's working on that.

File: resources/assets/js/components/Forms/TextArea.tsx (hunk 3 above)
Comment (xander-miller) [reply resources/assets/js/components/Forms/TextArea.tsx]: Yoni is working on this.

**Comment by xander-miller** [comment]
- [ ] Make sure scroll works after model activates.

File: resources/assets/js/components/JobBuilderImpact/JobImpactPreview.tsx [hunk 15]
```diff
@@ -0,0 +1,58 @@
...
+        data-c-border="bottom(thin, solid, black)"
+        data-c-font-size="h4"
+        data-c-font-weight="600"
+        data-c-margin="bottom(normal)"
+        data-c-padding="bottom(normal)"
+      >
+        Impact
+      </h4>
+      <p data-c-margin="bottom(normal)">
+        The Navigable Waters Act Renewal team is responsible for the
```
Comment (tristan-orourke) [inline resources/assets/js/components/JobBuilderImpact/JobImpactPreview.tsx]: Presumably these text sections are supposed to be replaced with {deptImpact}, {teamImpact} and {hireImpact}.

File: resources/assets/js/components/JobBuilderImpact/JobBuilderImpact.tsx (hunk 2 above)
Comment (tristan-orourke) [reply resources/assets/js/components/JobBuilderImpact/JobBuilderImpact.tsx]: We're already passing in `job` and getting the impact statements from there. Consistency suggests we should get the department from there as well.

**Comment by tristan-orourke** [comment]
> Handle Submit button isn't working, the modal is not appearing.
Switching from TextArea to TextAreaInput (the one designed for formik) seems to have fixed this.
